posix/posix.c

```diff
@@ -2686,45 +2686,68 @@ int posix_waitpid(pid_t child, int *status, int options)
 
 void posix_died(pid_t pid, int exit)
 {
-	process_info_t *pinfo, *ppinfo, *init, *cinfo;
+	process_info_t *pinfo, *ppinfo, *init, *cinfo, *zinfo, *zombies = NULL, *children = NULL;
+	int wait;
 
-	pinfo = pinfo_find(pid);
-	if (pinfo == NULL) {
-		return;
-	}
+	init = pinfo_find(1);
+	LIB_ASSERT_ALWAYS(init != NULL, "init not found");
 
-	posix_exit(pinfo, exit);
+	pinfo = pinfo_find(pid);
+	LIB_ASSERT_ALWAYS(pinfo != NULL, "pinfo not found, pid: %d", pid);
 
 	ppinfo = pinfo_find(pinfo->parent);
-	if (ppinfo != NULL) {
-		proc_lockSet(&ppinfo->lock);
-		LIST_REMOVE(&ppinfo->children, pinfo);
-		LIST_ADD(&ppinfo->zombies, pinfo);
-		proc_threadBroadcast(&ppinfo->wait);
-		proc_lockClear(&ppinfo->lock);
+	LIB_ASSERT_ALWAYS(pinfo != NULL, "parent not found, pid: %d, ppid: %d", pid, pinfo->parent);
 
-		init = pinfo_find(1);
-		if (init != NULL) {
-			proc_lockSet2(&pinfo->lock, &init->lock);
-			while (pinfo->children != NULL) {
-				cinfo = pinfo->children;
+	posix_exit(pinfo, exit);
+
+	proc_lockSet(&pinfo->lock);
+	/* Collect all children */
+	children = pinfo->children;
+	pinfo->children = NULL;
+	/* Collect all zombies */
+	zombies = pinfo->zombies;
+	pinfo->zombies = NULL;
+	proc_lockClear(&pinfo->lock);
 
-				/* Treat cinfo->parent as atomic, no need for lock */
-				cinfo->parent = 1;
+	/* Make a zombie, wakeup waitpid */
+	proc_lockSet(&ppinfo->lock);
+	LIST_REMOVE(&ppinfo->children, pinfo);
+	if (ppinfo != init) {
+		LIST_ADD(&ppinfo->zombies, pinfo);
+		wait = proc_threadBroadcast(&ppinfo->wait);
+	}
+	else {
+		LIST_ADD(&zombies, pinfo);
```

Comment:
I would add a comment here as I've missed at first that it handles the direct children of `init` to become zombies and leak.

Maybe we should also add comment in `init` process that the zombie reaping is being done in `posix_died` as I probably wouldn't think to look here?

Reply:
Ok, I'll add comment here